storage/block/block.go

```diff
@@ -31,95 +33,260 @@ import (
+	b.RLock()
+
 	if b.closed {
+		b.RUnlock()
 		return nil, errReadFromClosedBlock
 	}
+
 	b.ctx.DependsOn(blocker)
-	// If the block is not writable, and the segment is empty, it means
-	// there are no data encoded in this block, so we return a nil reader.
-	if b.segment.Head == nil && b.segment.Tail == nil {
-		return nil, nil
+
+	// Use an int64 to avoid needing a write lock for
+	// the high frequency stream method
+	atomic.StoreInt64(&b.accessedUnixNano, b.now().UnixNano())
+
+	// If the block retrieve ID is set then it must be retrieved
+	var (
+		stream xio.SegmentReader
+		err    error
+	)
+	if b.retriever != nil {
+		stream, err = b.retriever.Stream(b.retrieveID, b.startWithLock(), b)
+		if err != nil {
+			b.RUnlock()
+			return nil, err
+		}
+	} else {
+		stream = b.opts.SegmentReaderPool().Get()
+		// NB(r): We explicitly create a new segment to ensure references
+		// are taken to the bytes refs and to not finalize the bytes.
+		stream.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
 	}
-	s := b.opts.SegmentReaderPool().Get()
-	// NB(r): We explicitly create a new segment to ensure references
-	// are taken to the bytes refs and to not finalize the bytes.
-	s.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
-	blocker.RegisterFinalizer(context.FinalizerFn(s.Close))
-	return s, nil
+
+	if b.mergeTarget != nil {
+		var mergeStream xio.SegmentReader
+		mergeStream, err = b.mergeTarget.Stream(blocker)
+		if err != nil {
+			stream.Finalize()
+			return nil, err
+		}
+		// Return a lazily merged stream
+		// TODO(r): once merged reset this block with the contents of it
+		stream = newDatabaseMergedBlockReader(b.startWithLock(), stream, mergeStream, b.opts)
+	}
+
+	// Register the finalizer for the stream
+	blocker.RegisterFinalizer(stream)
+	b.RUnlock()
+
+	return stream, nil
+}
+
+func (b *dbBlock) IsRetrieved() bool {
+	b.RLock()
+	retrieved := b.retriever == nil
+	b.RUnlock()
+	return retrieved
+}
+
+func (b *dbBlock) Merge(other DatabaseBlock) {
+	b.Lock()
+	b.resetMergeTargetWithLock()
+	b.mergeTarget = other
+	b.Unlock()
 }
 
-// Reset resets the block start time and the encoder.
 func (b *dbBlock) Reset(start time.Time, segment ts.Segment) {
+	b.Lock()
+	defer b.Unlock()
+
+	if !b.closed {
+		b.ctx.Close()
+	}
+	b.ctx = b.opts.ContextPool().Get()
+	b.startUnixNano = start.UnixNano()
+	b.closed = false
+	b.resetMergeTargetWithLock()
+	atomic.StoreInt64(&b.accessedUnixNano, b.now().UnixNano())
+	b.resetSegmentWithLock(segment)
+}
+
+func (b *dbBlock) ResetRetrievable(
+	start time.Time,
+	retriever DatabaseShardBlockRetriever,
+	metadata RetrievableBlockMetadata,
+) {
+	b.Lock()
+	defer b.Unlock()
+
 	if !b.closed {
 		b.ctx.Close()
 	}
 	b.ctx = b.opts.ContextPool().Get()
-	b.start = start
+	b.startUnixNano = start.UnixNano()
 	b.closed = false
-	b.resetSegment(segment)
+	b.resetMergeTargetWithLock()
+	atomic.StoreInt64(&b.accessedUnixNano, b.now().UnixNano())
```

Comment:
This looks very similar to `Reset`, can we move all but the last line into a separate method and reuse it from `Reset` and `ResetRetrievable`?